File: server/src/routes/leaderboardRoutes.ts
import { Env, LeaderboardEntry, UserRecord } from "../types";
import { createJsonResponse, createErrorResponse } from "../utils/cors";
import { extractBearerToken } from "../utils/googleAuth";

/**
 * Handles GET /leaderboard/global
 */
export async function handleGlobalLeaderboard(_request: Request, env: Env): Promise<Response> {
  try {
    const { results } = await env.DB.prepare(
      "SELECT username, total_lifetime_points as points FROM users ORDER BY points DESC LIMIT 10"
    ).all();

    const leaderboard: LeaderboardEntry[] = results.map((row: any, index: number) => ({
      username: row.username,
      points: Number(row.points || 0),
      rank: index + 1,
    }));

    return createJsonResponse({ success: true, leaderboard });
  } catch (error: any) {
    return createErrorResponse(error.message || "Failed to fetch global leaderboard", 500);
  }
}

/**
 * Handles GET /leaderboard/friends
 */
export async function handleFriendsLeaderboard(request: Request, env: Env): Promise<Response> {
  try {
    const token = extractBearerToken(request);
    const user = await env.DB.prepare("SELECT id FROM users WHERE auth_token = ?").bind(token).first<UserRecord>();

    if (!user) {
      return createErrorResponse("Authentication required", 401);
    }

    const { results } = await env.DB.prepare(`
      SELECT username, total_lifetime_points as points FROM users WHERE id = ?
      UNION
      SELECT u.username, u.total_lifetime_points as points FROM friends f
      JOIN users u ON f.friend_id = u.id
      WHERE f.user_id = ?
      ORDER BY points DESC
    `)
      .bind(user.id, user.id)
      .all();

    const leaderboard: LeaderboardEntry[] = results.map((row: any, index: number) => ({
      username: row.username,
      points: Number(row.points || 0),
      rank: index + 1,
    }));

    return createJsonResponse({ success: true, leaderboard });
  } catch (error: any) {
    return createErrorResponse(error.message || "Failed to fetch friends leaderboard", 500);
  }
}

File: server/src/routes/leaderboard_routes.py
"""
Leaderboard Rankings Router for Prodo FastAPI Backend.

Provides endpoints for global focus leaderboards (`/leaderboard/global`)
and friends-only focus rankings (`/leaderboard/friends`).
"""

from typing import Optional
from fastapi import APIRouter, Header
from fastapi.responses import JSONResponse

from ..database import query_all, query_one
from .auth_routes import extract_bearer_token

router = APIRouter(prefix="/leaderboard", tags=["Leaderboard"])


@router.api_route("/global", methods=["GET", "POST"])
async def get_global_leaderboard():
    """
    Returns the top 50 users ranked by total focus XP.
    """
    sql = """
    SELECT username, xp AS points, total_lifetime_points
    FROM users
    ORDER BY xp DESC
    LIMIT 50
    """
    entries = query_all(sql)
    return {"success": True, "leaderboard": entries}


@router.api_route("/friends", methods=["GET", "POST"])
async def get_friends_leaderboard(authorization: Optional[str] = Header(None)):
    """
    Returns leaderboard rankings restricted to the authenticated user and their linked friends.
    """
    token = extract_bearer_token(authorization)
    if not token:
        # Fallback to global if no token provided
        return await get_global_leaderboard()

    user = query_one("SELECT * FROM users WHERE auth_token = ?", (token,))
    if not user:
        return await get_global_leaderboard()

    # Query friends + self rankings
    sql = """
    SELECT u.username, u.xp AS points
    FROM users u
    WHERE u.id = ? OR u.id IN (SELECT friend_id FROM friends WHERE user_id = ?)
    ORDER BY u.xp DESC
    """
    entries = query_all(sql, (user["id"], user["id"]))
    return {"success": True, "leaderboard": entries}
